File: 백준/Gold/2258. 정육점/정육점.py
from typing import List, Tuple

def solve_meat_problem(N: int, M: int, meats: List[Tuple[int, int]]) -> int:
    meats.sort(key=lambda x: (x[1], -x[0]))  # 가격 오름차순, 무게 내림차순 정렬
    
    weight = 0
    max_sum = 0
    now = 0
    
    i = 0
    for i in range(N):
        if weight >= M:
            break
        weight += meats[i][0]
        
        if meats[i][1] != now:
            now = meats[i][1]
            max_sum = meats[i][1]
        elif meats[i][1] == now:
            max_sum += meats[i][1]
            
    if weight >= M:
        for j in range(i, N):
            if meats[j][1] > now:
                if max_sum > meats[j][1]:
                    max_sum = meats[j][1]
                break
        return max_sum
    else:
        return -1

if __name__ == "__main__":
    N, M = map(int, input().split())
    meats = []
    for _ in range(N):
        weight, price = map(int, input().split())
        meats.append((weight, price))
    
    result = solve_meat_problem(N, M, meats)
    print(result)

File: 백준/Gold/2258. 정육점/정육점.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static class Meat implements Comparable<Meat> {
        long weight, price;

        public Meat(long w, long price) {
            this.weight = w;
            this.price = price;
        }

        @Override
        public int compareTo(Meat o) {
            if(this.price<o.price)
                return -1;
            else if(this.price>o.price)
                return 1;
            else
            {
                return Long.compare(o.weight, this.weight);
            }

        }

    }
        public static void main(String[] args) throws IOException {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

            StringTokenizer st = new StringTokenizer(br.readLine());

            int N = Integer.parseInt(st.nextToken());
            long M = Long.parseLong(st.nextToken());
            Meat[] meats = new Meat[N];
            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                meats[i] = new Meat(Long.parseLong(st.nextToken()), Long.parseLong(st.nextToken()));
            }

            Arrays.sort(meats);

            long weight = 0;

            int i = 0;
            long now= 0, maxSum = 0;
            for (; i < N; i++) {
                if (weight >= M)
                    break;
                weight += meats[i].weight;

                if (meats[i].price != now) {
                    now = meats[i].price;
                    maxSum = meats[i].price;
                } else if (meats[i].price == now) {
                    maxSum += meats[i].price;
                }
          
            }

            if (weight >= M){
                for(;i<N;i++) {
                    if(meats[i].price>now) {
                        if(maxSum>meats[i].price)
                            maxSum = meats[i].price;
                        break;
                    }
                }

                System.out.println(maxSum);
            }
            else
                System.out.println(-1);
        }


}
